### PR TITLE
Exposes get_object_stream using an AsyncRead interface

### PR DESCRIPTION
**Pull request by andrebsguedes**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1013,6 +1107,220 @@ pub extern "C" fn next_list_stream_chunk(
...
+        Some(runtime) => {
+            runtime.spawn(async move {
+                let read_op = async {
+                    let amount_to_read = size.min(amount);
+                    let mut bytes_read = 0;
+                    while amount_to_read > bytes_read {
+                        let n = wrapper.reader.read_buf(&mut slice).await?;
+
+                        if n == 0 {
+                            let eof = amount_to_read != bytes_read;
```
Comment (jfunstonRAI) [inline src/lib.rs]: Shouldn't this just be eof = true?

File: src/lib.rs [hunk 2]
```diff
@@ -1013,6 +1107,220 @@ pub extern "C" fn next_list_stream_chunk(
...
+impl Drop for ReadResponseGuard {
+    fn drop(&mut self) {
+        self.notify_on_drop()
+    }
+}
+
+unsafe impl Send for ReadResponseGuard {}
+
+pub struct GetStreamWrapper {
+    // stream: BoxStream<'static, Vec<Result<bytes::Bytes, anyhow::Error>>>
```
Comment (jfunstonRAI) [inline src/lib.rs]: Leftover comment?

File: src/lib.rs (hunk 1 above)
Comment (andrebsguedes) [reply src/lib.rs]: No, `read_buf` will also return 0 if the slice was completely filled, which is why we check if we read everything we requested and only if not we say it was an eof

File: src/lib.rs (hunk 2 above)
Comment (andrebsguedes) [reply src/lib.rs]: Yeah, will fix

File: src/lib.rs (hunk 1 above)
Comment (andrebsguedes) [reply src/lib.rs]: Taking another look, I think the while condition will never let us read 0 bytes and have `amount_to_read  == bytes_read` so eof will always be true as you said

**Review by jfunstonRAI — APPROVED** [review]

**Pull request merged**